### ISSUE TITLE
Code block tabs on mobile are not contained within a `tablist`

### ISSUE THREAD

**Issue by 36degrees** [alphagov/govuk-design-system#1653]
## Description of the issue

As [flagged by Lighthouse](https://wpt.speedcurve.com/lighthouse.php?test=210506_NX_164b84deeb72b9cda6092e7249246610&run=2) when run on the 'mobile' layout of the Design System, the HTML and Nunjucks 'tabs' are not contained within a `tablist`.

The tabs you see on mobile are different elements to the tabs you see on tablet and above, because on mobile the tabs act more like an accordion, and the elements need to be in a different position in the DOM.

The tablet / desktop tabs are correctly nested within a `ul role="tablist"`:

https://github.com/alphagov/govuk-design-system/blob/1fe2b366c1d13c457e6a210aaf54c3e7bc106774/views/partials/_example.njk#L54-L58

However, the mobile tabs are not:

https://github.com/alphagov/govuk-design-system/blob/1fe2b366c1d13c457e6a210aaf54c3e7bc106774/views/partials/_example.njk#L77

This means that the tabs are [not contained by their required parent element](https://web.dev/aria-required-parent/) – as [ARIA 1.1 says](https://www.w3.org/TR/wai-aria-1.1/#tab):

> Authors MUST ensure elements with role `tab` are contained in, or owned by, an element with the role `tablist`.

This fails [WCAG 2.1 1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html).

## Steps to reproduce the issue

In Chrome Developer Tools, inspect the accessibility tree for an HTML or Nunjucks 'tab', when the mobile layout is used.

## Actual vs expected behaviour

Expected: The tab should be contained within a tablist.
Actual: The tab is not contained within a tablist.

## Done when

- [ ] Issue fixed
- [ ] Issue removed from accessibility statement

**Comment by 36degrees** [comment]
Once this has been resolved, we'll need to [remove it from our accessibility statement](https://github.com/alphagov/govuk-design-system/pull/1655).

**Comment by vanitabarrett** [comment]
## Summary of my testing 
[View test results](https://docs.google.com/spreadsheets/d/1_0lC0v21yHEkoSzNg_iPkYvU_CzEohYM1CLbhkGGRfQ/edit#gid=0)

I tested 2 options against what is currently live (see below for more details). Both options work, and you get a clear expanded/collapsed state announced to the user. The only downside I can really see is that you lose the context of how many tabs there are within the section you’re on, for example: “HTML tab selected collapsed 1 of 2” vs “HTML button collapsed”.

From a look at other design systems, sites tend to use tabs or accordions but normally choose one across all screen sizes. From my observation, accordions tend to be the better choice as they work better on smaller screens as there is less likelihood of content wrapping or overflowing the page.

Some examples here:
- [Australian Government design system](https://designsystem.gov.au/components/breadcrumbs) uses tabs on mobile and desktop, with `role="tab"`
- [Elastic UI design system](https://elastic.github.io/eui/#/layout/bottom-bar) uses tabs on mobile and desktop, with `role="tab"`
- [United States of America design system](https://designsystem.digital.gov/components/button) uses an accordion with `aria-expanded` and `aria-controls`. The accordion uses `+` and `-` to indicate whether the section is opened or closed.

### Option 1: Remove `role=“tab”`
WIP branch: https://github.com/alphagov/govuk-design-system/commit/f3f8da9ac465834238bc1fe69df3b3c0751d7a56

The quickest change for us to make is to remove `role=“tab”` from the accordion markup on mobile. The accordion already has `aria-controls` and `aria-expanded` applied, which is similar to our own [accordion component markup](https://design-system.service.gov.uk/components/accordion/).

### Option 2: Remove `role=“tab”` and use buttons
WIP branch: https://github.com/alphagov/govuk-design-system/commit/a6753df585fe1660c1dd6d7230652b6557fa3a04

Removing `role=“tab”` means we are left with links that open and close each accordion section. These are probably more appropriate as buttons - the links do not really go anywhere, they open and close sections (which also matches our accordion component implementation).

### Option 3: Rewrite
Spike: https://github.com/alphagov/govuk-design-system/commit/4d935eb30f24a983e91199725d0bcf54ae7735cf

It may be simpler (for the user and for ourselves - the code for tabs/accordion is pretty complicated) to use one of accordion or tabs across all breakpoints. Guidance on [inclusive.components.design](https://inclusive-components.design/tabbed-interfaces/#responsivedesign) recommends not attempting to use both and says:

> Where there are very many tabs or the number of tabs are an unknown quantity, an accordion at all screen widths is a safe bet.

Our tabs are limited to 2 at the moment, so not sure if we can assume the same applies here, but it might be an interesting thing to test in user research.

## Note
A separate, but related issue I’ve noticed is that regardless of the implementation (accordion vs tabs; links vs buttons), we clutter up the page with a lot of interactive controls with the same text (“HTML” and “Nunjucks”). This might be noisy for users navigating by form controls or links or using speech recognition.

**Comment by 36degrees** [comment]
> It may be simpler (for the user and for ourselves - the code for tabs/accordion is pretty complicated) to use one of accordion or tabs across all breakpoints.

Happened to notice that gov.scot's design system uses accordions consistently at all breakpoints – https://designsystem.gov.scot/components/checkboxes/

**Comment by 36degrees** [comment]
I think option 2 generally makes sense, but we'll need to think about what the behaviour is on mobile when JavaScript fails or is not available.

The non-JS behaviour is for the tabs to fall back to being links. On desktop, this sort of makes sense as the links appear together at the top, and clicking them jumps to the relevant code block, but on mobile the links appear immediately before the individual code block, so don't offer a huge amount of value.

However, with option 2 we'll end up with non-functional buttons above each code block, which makes even less sense. Perhaps we could progressively enhance the HTML / Nunjucks text to turn it into a button using JS?

**Comment by hannalaakso** [comment]
Yes, I agree with Ollie both on preferring Option 2 and that we might need to progressively enhance the text to buttons with JS. Unless we think that having them as links when JS isn't running would have some value for navigating the page with screen readers or by tabbing.
